### PR TITLE
Security fix 4 - Simplify node time and network-adjusted time warning logic

### PR DESCRIPTION
**Pull request by Naviabheeman**
Port fix of vulnerability fix of [Disclosure of netsplit due to timestamp adjustment](https://bitcoincore.org/en/2024/07/03/disclose-timestamp-overflow/) which was changed later by PR 29623(Simplify network-adjusted time warning logic)

As this PR removes the `timeoffset` previously defined in each node, the notion of 'adjusted time' is also removed. So a unified `std::chrono::system_clock` is used to get the time in GetTime, GetAdjustedTime, GetSystemTimeInSeconds, GetTimeMillis and GetTimeMicros. 

- `GetAdjustedTime` and `GetTime` are now the same. They return the mockable system time. 
- `GetSystemTimeInSeconds`  does not allow mocktime. This is essential as the p2p layer depends heavily on the connection and message time. Any change to this time due to mocking results in unpredictable behaviour in functional tests.

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -203,7 +203,6 @@ add_library(server
         script/descriptor.cpp
         script/sigcache.cpp
         script/ismine.cpp
-        timedata.cpp
```
Comment (azuchi) [inline src/CMakeLists.txt:206]: Do you need to add `timeoffsets.cpp`?

**Review by azuchi — CHANGES_REQUESTED** [review]
Isn't the `nTimeOffset` field in classes `CNode` and `CNodeStats` also unnecessary?

File: src/net_processing.cpp [hunk 2]
```diff
@@ -23,6 +23,7 @@
 #include <random.h>
 #include <reverse_iterator.h>
 #include <scheduler.h>
+#include <timeoffsets.h>
```
Comment (azuchi) [inline src/net_processing.cpp:26]: Are the following changes to Bitcoin unnecessary?

https://github.com/bitcoin/bitcoin/pull/29623/files#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R1801

File: src/rpc/net.cpp [hunk 3]
```diff
@@ -426,7 +429,6 @@ static UniValue getnetworkinfo(const JSONRPCRequest& request)
             "  \"protocolversion\": xxxxx,              (numeric) the protocol version\n"
             "  \"localservices\": \"xxxxxxxxxxxxxxxx\", (string) the services we offer to the network\n"
             "  \"localrelay\": true|false,              (bool) true if transaction relay is requested from peers\n"
-            "  \"timeoffset\": xxxxx,                   (numeric) the time offset\n"
```
Comment (azuchi) [inline src/rpc/net.cpp:429]: It doesn't seem to have been removed in Bitcoin.

File: src/test/timedata_tests.cpp [hunk 4]
```diff
@@ -3,7 +3,6 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 //
-#include <timedata.h>
```
Comment (azuchi) [inline src/test/timedata_tests.cpp:6]: Bitcoin has removed this test file itself. Why keep this?

File: test/functional/test_framework/mininode.py [hunk 5]
```diff
@@ -299,6 +299,7 @@ def on_reject(self, message): pass
     def on_sendcmpct(self, message): pass
     def on_sendheaders(self, message): pass
     def on_tx(self, message): pass
+    def on_notfound(self, message): pass
```
Comment (azuchi) [inline test/functional/test_framework/mininode.py:302]: Is this change related to this PR?

**Review by azuchi — CHANGES_REQUESTED** [review]

**Comment by Naviabheeman** [comment]
> Isn't the `nTimeOffset` field in classes `CNode` and `CNodeStats` also unnecessary?

`nTimeOffset` field  in `CNode` class has the time offset for each peer. I was not able to remove it.

File: src/net_processing.cpp [hunk 6]
```diff
@@ -1863,9 +1864,13 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
                   pfrom->nStartingHeight, addrMe.ToString(), pfrom->GetId(),
                   remoteAddr);
 
-        int64_t nTimeOffset = nTime - GetTime();
-        pfrom->nTimeOffset = nTimeOffset;
-        AddTimeData(pfrom->addr, nTimeOffset);
+        pfrom->nTimeOffset = nTime - GetTime();
```
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply src/net_processing.cpp:1867]: In Bitcoin there are lots of readable user friendly names. Each is actually just a typedef
`using NodeSeconds = std::chrono::time_point<NodeClock, std::chrono::seconds>;`
In our code the long qualified names are used. And the interface is not changed i.e. GetTime(), GetSystemTime() etc are retained.

File: src/qt/rpcconsole.cpp [hunk 7]
```diff
@@ -1120,7 +1120,6 @@ void RPCConsole::updateNodeDetail(const CNodeCombinedStats *stats)
     ui->peerPingTime->setText(GUIUtil::formatPingTime(stats->nodeStats.dPingTime));
     ui->peerPingWait->setText(GUIUtil::formatPingTime(stats->nodeStats.dPingWait));
     ui->peerMinPing->setText(GUIUtil::formatPingTime(stats->nodeStats.dMinPing));
-    ui->timeoffset->setText(GUIUtil::formatTimeOffset(stats->nodeStats.nTimeOffset));
     ui->peerVersion->setText(QString("%1").arg(QString::number(stats->nodeStats.nVersion)));
```
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply src/qt/rpcconsole.cpp:1123]: It should be there. I have missed it.

File: src/net_processing.cpp (hunk 2 above)
Comment (Naviabheeman) [reply src/net_processing.cpp:26]: 'CNodeStateStats' has more information in Bitcoin than in Tapyrus. It was changed while refactoring for Peer manager class. In out code time offset is still in 'CNodeStats'

File: src/rpc/net.cpp (hunk 3 above)
Comment (Naviabheeman) [reply src/rpc/net.cpp:429]: Yes, this is a little different now. Earlier, each node maintains its own time offset from to the median block time. So this information was available globally and reported in network info. Now it is maintained by the node for each peer and reported only in peerinfo rpc.

File: test/functional/test_framework/mininode.py (hunk 5 above)
Comment (Naviabheeman) [reply test/functional/test_framework/mininode.py:302]: no, this code is not necessary. will remove it.

**Comment by azuchi** [comment]
> > Isn't the `nTimeOffset` field in classes `CNode` and `CNodeStats` also unnecessary?
> 
> `nTimeOffset` field in `CNode` class has the time offset for each peer. I was not able to remove it.

but, `nTimeOffset` in `CNodeStats` seems not referenced anywhere.

**Comment by Naviabheeman** [comment]
> but, `nTimeOffset` in `CNodeStats` seems not referenced anywhere.

oh! yes I'm removing it.

**Review by azuchi — APPROVED** [review]
